charms/slurmctld/src/integrations.py

```
@@ -0,0 +1,128 @@
+
+_logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class ControllerPeerAppData:
+    """Data provided by the primary/leader Slurm controller, `slurmctld`.
+
+    Attributes:
+        cluster_name: The unique name of this cluster.
+
+    Warnings:
+        - `cluster_name` can only be set when the peer integration is created. The cluster name
+           becomes permanent after the `SlurmctldPeerConnectedEvent` is handled.
+    """
+
+    cluster_name: str = ""
+
+
+class SlurmctldPeerConnectedEvent(ops.RelationEvent):
+    """Event emitted when `slurmctld` is connected to the peer integration."""
+
+
+class _SlurmctldPeerEvents(ops.CharmEvents):
+    """`slurmctld` peer events."""
+
+    slurmctld_peer_connected = ops.EventSource(SlurmctldPeerConnectedEvent)
+
+
+class SlurmctldPeer(Interface):
+    """Integration interface implementation for `slurmctld` peers."""
+
+    on = _SlurmctldPeerEvents()  # type: ignore
+
+    def __init__(self, charm: ops.CharmBase, integration_name: str) -> None:
+        super().__init__(charm, integration_name)
+
+        self.charm.framework.observe(
+            self.charm.on[self._integration_name].relation_created,
+            self._on_relation_created,
+        )
+
+    @leader
+    def _on_relation_created(self, event: ops.RelationCreatedEvent) -> None:
+        """Handle when `slurmctld` peer integration is first established."""
+        self.on.slurmctld_peer_connected.emit(event.relation)
+
+    @leader
+    def set_controller_peer_app_data(self, data: ControllerPeerAppData, /) -> None:
+        """Set the controller peer data in the `slurmctld-peer` application databag.
+
+        Args:
+            data: Controller peer data to set in the peer integration's application databag.
+
+        Warnings:
+            - Only the `slurmctld` application leader can set controller peer app data.
+        """
+        integration = self.get_integration()
+        if not integration:
```

I think we always expect `get_integration()` to find an integration so it would make sense for it to raise an exception if it doesn't, rather than return `None`. Then we could have something like:

```python
try:
    integration = self.get_integration()
    integration.save(data, self.app)
    _logger.info("set data")
except:
    _logger.info("did not set data")
```

Forgiveness rather than permission to do `integration.save()`, and similar for the `integration.load()` below. This would require an `hpc-libs` change but I think we should consider it.

---

I'd consider that out of scope for this PR since this PR is primarily focused on introducing the local `slurm_ops` changes, but I needed the initial peer interface updates so that the integration tests would work. 

That said, the reason why I prefer the `None` check is because typically comparing against `None` is faster than exception handling:

```python
import logging
import os
import sys
import timeit

logger = logging.getLogger()

fout = open(os.devnull, "w")
sys.stderr = fout


def lookup_check_if_none():
    data = {1: "some cool integration"}
    integration = data.get(2)
    if integration is None:
        logger.error("no integration found :(")


def lookup_handle_key_error():
    data = {1: "some cool integration"}
    try:
        _ = data[2]
    except KeyError:
        logger.error("no integration found :(")


if __name__ == "__main__":
    print(
        f"`lookup_check_if_none` ttc: {timeit.timeit(lambda: lookup_check_if_none(), number=10000)}"
    )
    print(
        f"`lookup_handle_key_error` ttc: {timeit.timeit(lambda: lookup_handle_key_error(), number=10000)}"
    )

# Output:
# `lookup_check_if_none` ttc: 0.0709606949967565
# `lookup_handle_key_error` ttc: 0.07322793199273292
```

Also, you can call `get_integration` whenever in the charm, we just only call it when there's a high likelihood that the integration does exist.

---

The `get_integration(...)` method is effectively a forward for `self.charm.model.get_relation(...)` where the integration name is already set to the integration name the interface class was initialized with. `get_relation(...)` doesn't return an error if the integration is not found, so we'd need to decide if we want to make `get_integration` raise an exception if `get_relation(...)` returns `None`. We'd would likely need to add a custom exception like`"IntegrationNotFoundError` or something similar.

---

Either way, since we've identified these changes as out of scope for this PR, let's continue this conversation in a new PR which will introduce interface changes in `hpc_libs`.